web/src/saved_replies.ts

```diff
@@ -0,0 +1,161 @@
+    const $submit_button = $("#add-new-saved-reply-modal .dialog_submit_button");
+
+    $submit_button.prop("disabled", true);
+    if (title && content) {
+        $submit_button.prop("disabled", false);
+    }
+}
+
+function saved_reply_modal_post_render(): void {
+    $("#add-new-saved-reply-modal").on("input", "input,textarea", update_submit_button_state);
+
+    $("#add-new-saved-reply-modal").on("keypress", "input", (event) => {
+        if (keydown_util.is_enter_event(event)) {
+            event.preventDefault();
+            submit_create_saved_reply_form();
+        }
+    });
+}
+
+function item_click_callback(
+    event: JQuery.ClickEvent,
+    dropdown: tippy.Instance,
+    widget: dropdown_widget.DropdownWidget,
+): void {
+    event.preventDefault();
+    event.stopPropagation();
+
+    if ($(event.target).closest(".delete-saved-reply").length) {
+        const id = $(event.currentTarget).attr("data-unique-id");
+        if (id === undefined) {
+            return;
+        }
+        const saved_reply_id = Number.parseInt(id, 10);
+        void channel.del({
+            url: "/json/saved_replies/" + saved_reply_id,
+            success() {
+                const options = get_options_for_dropdown_widget();
+                widget.list_widget?.replace_list_data(
+                    options.filter((option) => option.unique_id !== saved_reply_id),
+                );
+            },
+        });
+        return;
+    }
+
+    dropdown.hide();
+    const current_value = widget.current_value;
+    if (current_value === ADD_SAVED_REPLY_OPTION_ID) {
+        dialog_widget.launch({
+            html_heading: $t_html({defaultMessage: "Add a new saved reply"}),
+            html_body: render_add_saved_reply_modal(),
+            html_submit_button: $t_html({defaultMessage: "Save"}),
+            id: "add-new-saved-reply-modal",
+            update_submit_disabled_state_on_change: true,
+            on_click: submit_create_saved_reply_form,
+            post_render: saved_reply_modal_post_render,
+        });
+    } else {
+        const content = my_saved_replies.find(
+            (saved_reply) => saved_reply.id === current_value,
+        )?.content;
+        assert(content !== undefined);
+        const $textarea = $<HTMLTextAreaElement>("textarea#compose-textarea");
+        compose_ui.insert_and_scroll_into_view(content, $textarea);
+    }
+}
+
+export const initialize = (params: StateData["saved_replies"]): void => {
+    set_saved_replies(params.saved_replies);
+
+    new dropdown_widget.DropdownWidget({
+        widget_name: "saved_replies",
+        get_options: get_options_for_dropdown_widget,
+        item_click_callback,
+        $events_container: $("body"),
+        unique_id_type: dropdown_widget.DataTypes.NUMBER,
+        focus_target_on_hidden: false,
+        prefer_top_start_placement: true,
+        tippy_props: {
+            offset: [-10, 5],
```

Comment:
Can you add a comment documenting these numbers?

Reply:
They were copied from compose recipient dropdown. Changed them now and added a comment:
```
// Using -100 as x offset makes saved reply icon be in the center
// of the dropdown widget and 5 as y offset is what we use in compose
// recipient dropdown widget.
```